File: BeadSnap/BeadSnap/ViewModels/LibraryViewModel.swift
import SwiftUI
import Combine

enum LibrarySortOrder: String, CaseIterable, Identifiable {
    case title, difficulty, newest
    var id: String { rawValue }
    var displayName: String {
        switch self {
        case .title:      return "Name"
        case .difficulty: return "Difficulty"
        case .newest:     return "Newest First"
        }
    }
    var systemImage: String {
        switch self {
        case .title:      return "textformat.abc"
        case .difficulty: return "chart.bar"
        case .newest:     return "clock"
        }
    }
}

@MainActor
final class LibraryViewModel: ObservableObject {
    @Published var selectedCategory: PatternCategory? = nil
    @Published var searchQuery = ""
    @Published var sortOrder: LibrarySortOrder = .title
    @Published var patterns: [FusePattern] = []
    @Published private(set) var categoryCounts: [PatternCategory: Int] = [:]

    private let store = PatternStore.shared
    private var cancellables = Set<AnyCancellable>()

    init() {
        // Debounce only keystrokes — chip taps, sort changes, and store updates
        // re-filter immediately (matches Android's query-only debounce)
        let debouncedQuery = $searchQuery
            .removeDuplicates()
            .debounce(for: .milliseconds(80), scheduler: DispatchQueue.main)

        store.$systemPatterns
            .combineLatest(store.$userPatterns, $selectedCategory, debouncedQuery)
            .combineLatest($sortOrder)
            .sink { [weak self] combined, sort in
                let (system, user, category, query) = combined
                self?.filter(system: system, user: user, category: category, query: query, sort: sort)
            }
            .store(in: &cancellables)

        store.$systemPatterns
            .combineLatest(store.$userPatterns)
            .sink { [weak self] system, user in
                var counts: [PatternCategory: Int] = [:]
                for p in system { counts[p.category, default: 0] += 1 }
                for p in user { counts[p.category, default: 0] += 1 }
                self?.categoryCounts = counts
            }
            .store(in: &cancellables)
    }

    private func filter(
        system: [FusePattern], user: [FusePattern],
        category: PatternCategory?, query: String,
        sort: LibrarySortOrder
    ) {
        let all = system + user
        var result = all
        if let cat = category { result = result.filter { $0.category == cat } }
        let q = query.lowercased().trimmingCharacters(in: .whitespaces)
        if !q.isEmpty {
            result = result.filter {
                $0.title.lowercased().contains(q) ||
                $0.tags.contains { $0.lowercased().contains(q) }
            }
        }

        switch sort {
        case .title:
            result.sort { $0.title.lowercased() < $1.title.lowercased() }
        case .difficulty:
            let order: [Difficulty] = [.easy, .medium, .hard]
            result.sort {
                (order.firstIndex(of: $0.difficulty) ?? 0) <
                (order.firstIndex(of: $1.difficulty) ?? 0)
            }
        case .newest:
            // User patterns come last in `all`; reverse so user patterns appear first
            result = result.reversed()
        }

        patterns = result
    }

    func count(for category: PatternCategory) -> Int {
        categoryCounts[category] ?? 0
    }
}

File: BeadSnapAndroid/app/src/main/kotlin/com/beadsnap/app/ui/screens/library/LibraryViewModel.kt
package com.beadsnap.app.ui.screens.library

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.beadsnap.app.data.model.Difficulty
import com.beadsnap.app.data.model.FusePattern
import com.beadsnap.app.data.model.PatternCategory
import com.beadsnap.app.data.store.PatternStore
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.stateIn

enum class LibrarySortOrder(val displayName: String) {
    TITLE("Name"),
    DIFFICULTY("Difficulty"),
    NEWEST("Newest First")
}

@OptIn(FlowPreview::class, ExperimentalCoroutinesApi::class)
class LibraryViewModel(private val store: PatternStore) : ViewModel() {

    private val _selectedCategory = MutableStateFlow<PatternCategory?>(null)
    val selectedCategory: StateFlow<PatternCategory?> = _selectedCategory.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _sortOrder = MutableStateFlow(LibrarySortOrder.TITLE)
    val sortOrder: StateFlow<LibrarySortOrder> = _sortOrder.asStateFlow()

    val patterns: StateFlow<List<FusePattern>> = combine(
        store.systemPatterns,
        store.userPatterns,
        _selectedCategory,
        _searchQuery.debounce(80),
        _sortOrder
    ) { system, user, category, query, sort ->
        filter(system + user, category, query, sort)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyList())

    val categoryCounts: StateFlow<Map<PatternCategory, Int>> = combine(
        store.systemPatterns, store.userPatterns
    ) { system, user ->
        (system + user).groupingBy { it.category }.eachCount()
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyMap())

    fun setCategory(cat: PatternCategory?) { _selectedCategory.value = cat }
    fun setQuery(q: String) { _searchQuery.value = q }
    fun setSortOrder(order: LibrarySortOrder) { _sortOrder.value = order }

    private fun filter(
        all: List<FusePattern>,
        category: PatternCategory?,
        query: String,
        sort: LibrarySortOrder
    ): List<FusePattern> {
        var result = if (category != null) all.filter { it.category == category } else all
        val q = query.trim().lowercase()
        if (q.isNotEmpty()) {
            result = result.filter {
                it.title.lowercase().contains(q) || it.tags.any { t -> t.lowercase().contains(q) }
            }
        }
        return when (sort) {
            LibrarySortOrder.TITLE      -> result.sortedBy { it.title.lowercase() }
            LibrarySortOrder.DIFFICULTY -> result.sortedBy { difficultyOrder(it.difficulty) }
            LibrarySortOrder.NEWEST     -> result.reversed()
        }
    }

    private fun difficultyOrder(d: Difficulty) = when (d) {
        Difficulty.easy -> 0; Difficulty.medium -> 1; Difficulty.hard -> 2
    }
}
